Robust.Client/Audio/Midi/MidiRenderer.cs

```
@@ -0,0 +1,383 @@
+            const int ch = 0;
+
+            try
+            {
+                switch (midiEvent.Type)
+                {
+                    // NoteOff
+                    case 128:
+                        if (_notesPlaying.Contains(midiEvent.Key))
+                        {
+                            _synth.NoteOff(ch, midiEvent.Key);
+                            _notesPlaying.Remove(midiEvent.Key);
+                        }
+
+                        break;
+                    // NoteOn
+                    case 144:
+                        // NoteOn with 0 velocity is the same as NoteOff
+                        if (midiEvent.Velocity == 0)
+                        {
+                            if (_notesPlaying.Contains(midiEvent.Key))
+                            {
+                                _synth.NoteOff(ch, midiEvent.Key);
+                                _notesPlaying.Remove(midiEvent.Key);
+                            }
+                        }
+                        else
+                        {
+                            // If we're at the limit of notes at once, we drop this one.
+                            if (_notesPlaying.Count >= NoteLimit)
+                                return;
+
+                            _synth.NoteOn(ch, midiEvent.Key, midiEvent.Velocity);
+                            if (!_notesPlaying.Contains(midiEvent.Key))
+                                _notesPlaying.Add(midiEvent.Key);
+                        }
+
+                        break;
+                    default:
+                        if (_notesPlaying.Contains(midiEvent.Key))
+                        {
+                            _synth.NoteOff(ch, midiEvent.Key);
+                            _notesPlaying.Remove(midiEvent.Key);
+                        }
+                        break;
+                }
+            }
+            catch (FluidSynthInteropException interopException)
```

Why are you ignoring exceptions here?
If you have a good reason, please do elaborate in a comment.

---

At first I was ignoring all exceptions because trying to send a NoteOff event for a key that's not on threw an exception... However, after adding the list of played notes this isn't the case anymore. Furthermore, I think a fluidsynth exception shouldn't crash the game, so I think I'll add a log message here instead of throwing the exception. Is that fine?

---

Yeah, that's fine.